Hooks

shaladdle opened the pull request:
## `pre-push`

1. Check that the working copy is clean
2. Check for `multirust`
3. If `multirust` is present, run tests on various toolchains
4. If it's not present, just run tests with current toolchain

## `pre-commit`

1. Check for non-ascii filenames
2. Check for bad whitespace in files
3. Check for existence of rustfmt + shasum
4. If they are present, check that files being committed are properly formatted

shaladdle commented:
I signed it!

tikue commented on hooks/pre-push:
```diff
@@ -10,25 +33,80 @@ GREEN='\033[0;32m'
-	exit 1
+	printf "${SUCCESS}\n"
+else
+	printf "${FAILURE}\n"
```
Looks like this printf should go in the else block

tikue commented:
So, build on all 3 trains, but only test on nightly?

tikue commented on hooks/pre-push:
```diff
@@ -10,25 +33,80 @@ GREEN='\033[0;32m'
-	TEST_RESULT=1
+	printf "${SUCCESS}\n"
+
+	# No need to build on nightly, the tests will do that
```
We do need to build on nightly, because there are dev dependencies only built when testing, so there are failure modes only when building, e.g. if an `extern crate` isn't `#[cfg(test)]`ed but it's listed under `[dev-dependencies]`

shaladdle commented:
- Now building on nightly too
- Fixed that printf
- Fixed some minor other things

[sample pre-push output](https://www.dropbox.com/s/k20z2gwigzuydis/Screen%20Shot%202016-02-18%20at%201.45.20%20AM.png?dl=0), [sample pre-commit output](https://www.dropbox.com/s/4bsty93u2m7z13t/Screen%20Shot%202016-02-18%20at%201.49.23%20AM.png?dl=0)
[me writing bash](http://img1.rnkr-static.com/user_node_img/50013/1000240969/870/doing-science-photo-u1.jpg)

tikue commented on hooks/pre-push:
```diff
@@ -10,30 +33,95 @@ GREEN='\033[0;32m'
-	printf "${GREEN}ok${NC}"
-fi
-printf "\n"
+	printf "${FAILURE}\n"
```
I don't know if success/failure should be printed when checking for the toolchain, since we proceed in either case. Isn't warning enough?

tikue replied:
"ok" is good for success, but maybe just not "FAILURE" when a toolchain is missing

tikue commented:
Works great for me! One question -- I'm a bit concerned about having to remember to `cp hooks/* .git/hooks` whenever there are changes. Could we instead have the thing in `.git/hooks` delegate to the `hooks/` directory?

shaladdle commented:
Changed to print NOT FOUND instead of FAILURE in the case that multirust isn't found. If you want the scripts to auto update, just symilnk instead of copying.
